rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,189 @@
+    "event_timestamp": "2019-10-23T04:00:00.000Z",
+    "event_type": "NETWORK_HTTP",
+    "product_name": "Acme Proxy",
+    "vendor_name": "Acme"
+  },
+  "network": {
+    "http": {
+      "method": "GET",
+      "response_code": 200
+    }
+  },
+  "principal": {
+    "hostname": "host0",
+    "ip": "10.1.2.3",
+    "port": 60000
+  },
+  "target": {
+    "hostname": "www.altostrat.com",
+    "ip": "198.51.100.68",
+    "port": 443,
+    "url": "www.altostrat.com/images/logo.png"
+  }
+}
+```
+
+Events can be batched and sent in a JSON object that also includes the
+`customer_id`:
+
+```json
+{
+"customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+"events": [...]
+}
+```
+
+UDM events can be sumbitted as either JSON or Proto3. We will submit using JSON
+as this fits in easier with Vectors data model. The newly developed schema
+support for Vector will be used to ensure the JSON is formed according to the
+[Chronicle requirements](https://cloud.google.com/chronicle/docs/unified-data-model/format-events-as-udm).
+
+Vector has no internal knowledge of SIEM events so it will be up to the user to
+ensure the event is well formed - either the source will receive the data in the
+correct structure, or it will be transformed into such before reaching the sink.
+
+#### Unstructured log data events
+
+Unstructured log events are events in plain text with no strict schema attached.
+
+```json
+{
+  "log_text": "26-Feb-2019 13:35:02.187 client 10.120.20.32#4238: query: altostrat.com IN A + (203.0.113.102)",
+  "ts_epoch_microseconds": 1551188102187000
+}
+```
+
+Events are batched and sent in a JSON object that includes the `customer_id` and
+a `log_type`:
+
+```json
+{
+  "customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable.
+
+Another field `remove_log_type` allows the field used in the template to be
+removed from the event.
+
+The encoding of the message (text or json) can be set via an `encoding` field.
+
+### Implementation
+
+Both event types require different batching strategies.
+
+#### UDM
+
+UDM can use a simple `BatchedHttpSink` with a `JsonArrayBuffer` to perform the
+batching.
+
+#### Unstructured log events
+
+Since log events are batched by `log_type` we will need to partition the
+batches on the value of the `log_type` field using a `PartitionHttpSink`.
+
+## Rationale
+
+There is significant user demand for sending data into Google Chronicle.
+
+## Drawbacks
+
+Adding a new sink does place an additional maintenance burden on the team. In
+particular since Chronicle is a third party service, developing integration
+tests to ensure the sink continues to work correctly is burdensome.
```

Comment:
Any thoughts on how easy it will be to add those integration tests? Any containers that mimic the API of this service? Do we need to stub our own mock API?

Reply:
I haven't found any containers, so it does look like we would need to stub our own mock API for this.